[PoS] sys/flash_utils: helpers to store data in flash

maribu opened the pull request:
### Contribution description

This helpers that allow storing, accessing, and working with data in flash that works for both classical Harvard architectures (which do not map flash also into the address space) as well as modern Harvard architectures and von-Neumann architectures.

With this, `examples/default` again runs on the Arduino Uno / Nano. Since this board is still the "entry kit" for many people to embedded hardware, it would be nice to support it with our default example.

### Testing procedure

`examples/default` should run and work on ATmega boards (especially ATmega328P and ATmega32U4 based boards) as well on all other boards now.

### Issues/PRs references

None

maribu commented on sys/include/flash_utils.h:
```diff
@@ -0,0 +1,329 @@
+    return ram;
+}
+
+#define flash_printf(...) printf(__VA_ARGS__)
+#define FLASH_PRINTF(...) printf(__VA_ARGS__)
+#define flash_snprintf(...) snprintf(__VA_ARGS__)
+#define FLASH_SNPRINTF(...) printf(__VA_ARGS__)
```
Here is what I tried to combine the two cases into a single macro:

```C
#define flash_puts(str) \
    puts_P( _Generic(str, \
        __flash const char *: str, \
        char *: __extension__({static __flash const char __c[] = (str); &__c[0];})))
```

The alternative would be to only provide one set of functions and let the caller do something like:

```C
flash_puts(TO_FLASH("Hello world"));
```

That is what the avr-libc is doing with the `PSTR()` macro.

kaspar030 reviewed (changes requested):
I don't think this is worth the maintenance overhead. It makes code more unique, less readable, less idiomatic, less understandable, for all platforms. All to save RAM on low-end boards like ATmega328P and ATmega32U4.

There was an understanding that RIOT does support Atmegas somewhat, but won't compromise too much for that support. This PR crosses that line.

There are many reasons:

- `_flash` et all, where do we use that? Are we gonna request new PRs to use that and flash_utils where appropriate? I guess we do agree that those *at least* decrease readability and increase the uniqueness of our codebase, which is rarely a good thing. And this forces a specific coding style just for the benefit of an obsolete architecture...

- the mapped functions (strlen, strcmp, ...) are difficult enough to get right with the combination of multiple libcs, multiple compilers, multiple intrinsics (libgcc, compiler-rt), multiple optimization levels. Adding another layer on top is opening a can of worms.

- macros. they seem to creep in lately. but they suck. XFAs just doesn't work without, but are pretty contained, there's no other way, and they (potentially) provide huge benefits. In rare cases, macros increase readability, at a cost. There's "standard macros" (`ARRAY_SIZE`) which _almost_ go by as idiomatic C. But wrapping `puts()` and `printf()`? please no! `PRIsflash` is also way too special for my taste.

maribu commented:
I think this is ready for review now. I could split out the actual users into separate PRs, though.

> `_flash` et all, where do we use that?

Exactly where it is introduced now:
1. shell command names and their help strings
    - this is fully transparent and only people actually working on the shell interpreter will notice
2. unit string (phydat) and sensor / actuator strings (SAUL)
    - this is mostly transparent, as except for the shell command I expect few users of those strings
3. In `ps`
    - this doesn't safe as much RAM as the other two, but it is fully transparent and `ps.c` is as "finished" as code can be

Right now we loose the AVR platform on these three items. This PR as is wins the platform back and makes the main features usable again on the most common board. There will be a few bytes here and there that could also be safed, but this really was the major chunk that actually makes a difference.

> the mapped functions (strlen, strcmp, ...) are difficult enough to get right with the combination of multiple libcs, multiple compilers, multiple intrinsics (libgcc, compiler-rt), multiple optimization levels. Adding another layer on top is opening a can of worms.

We have `sys/fmt` which is more pain for less gain.

Also, this doesn't add any new functions. Right now the way to go is:

``` C
#ifdef __AVR__
    puts_P(PSTR("Hello, world!"));
#else
    puts("Hello, world!");
#endif
```

With this PR users can simplify their code to:

``` C
    flash_puts(TO_FLASH("Hello, world!"));
```

It is just about making the dispatching between the `printf()` functions easier.

> macros. they seem to creep in lately. but they suck.

Note that the macros as simple as

```C
#define flash_printf printf_P
```
or
```C
#define flash_printf printf
```

That certainly is better than


``` C
#ifdef __AVR__
    puts_P(PSTR("Hello, world!"));
#else
    puts("Hello, world!");
#endif
```

But back to the main point: Those old shitty ATmega Arduino boards are still ***the*** first board people buy when they get in touch with microcontroller development. We have actually quite a lot of PRs from "outside contributors", e.g.

- https://github.com/RIOT-OS/RIOT/pull/15494
- https://github.com/RIOT-OS/RIOT/pull/15712
- https://github.com/RIOT-OS/RIOT/pull/15758
- https://github.com/RIOT-OS/RIOT/pull/15998
- https://github.com/RIOT-OS/RIOT/pull/16288
- https://github.com/RIOT-OS/RIOT/pull/16587
- https://github.com/RIOT-OS/RIOT/pull/17387
- https://github.com/RIOT-OS/RIOT/pull/18245

If we could keep them motivated to stick with RIOT and contribute more, that would be immensely helpful. But right now the default example greets them with a linking error. We can fix that.

maribu commented:
OK, so `flash_utils` is C only. This means it may not be included in any non-internal header. But this is not much of an issue, as this should be contained well anyway.

I have to rework the SAUL / phydat changes, though.

maribu commented:
This is now self-contained also for the `<foo>_to_str()` functions in SAUL and phydat, so that C++ code never will indirectly include `flash_utils.h`. However, this deprecates the `phydat_unit_to_str()` and `saul_class_to_str()`, as those use a shared in-RAM buffer for AVR that doesn't work well with concurrent use or even `printf("%s, %s", phydat_unit_to_str(UNIT_NONE), phydat_unit_to_str(UNIT_UNDEF));`. Hence the API change label.

However, users will want to either print those strings (that is now covered) or write that string into some buffer for sending it out via network (that is now also covered) anyway. And non-AVR users actually can keep using the function without fear or overhead.

benpicco commented on sys/phydat/phydat_json.c:
```diff
@@ -84,13 +85,11 @@ size_t phydat_to_json(const phydat_t *data, size_t dim, char *buf)
         }
         /* add unit */
-        memcpy(&buf[pos], "\"u\":\"", 5);
+        flash_memcpy(&buf[pos], TO_FLASH("\"u\":\""), 5);
```
Isn't it implicit that the second argument of `flash_memcpy` must always be in the flash?

maribu replied:
Yes. The issue is that string literals are not placed in flash. Ideally a `_Generic` would be used to just force string literals to flash, but also except `__flash const char *` pointers, and fail to compile for `char *` or `const char *`.

However, there seems to be no way to actually test for string literals :/

benpicco commented on sys/phydat/phydat_str.c:
```diff
@@ -18,149 +18,229 @@
         assert(dim == 3);
-        printf("\t%02d:%02d:%02d\n",
-               data->val[2], data->val[1], data->val[0]);
+        flash_printf(TO_FLASH("\t%02d:%02d:%02d\n"),
```
Could we enforce the first argument of `flash_printf` to always be in the flash? (so the `TO_FLASH` wrapper is not needed to be manually added at every call site)

benpicco replied:
How about

```C
#define flash_str_printf(s, ...) flash_printf(TO_FLASH(s), __VA_ARGS__)
```

benpicco commented:
IIRC you mentioned you had an idea on how to get rid of the `TO_FLASH` in `flash_printf(TO_FLASH())` - that would make this a lot more palatable

maribu commented:
> IIRC you mentioned you had an idea on how to get rid of the `TO_FLASH` in `flash_printf(TO_FLASH())` - that would make this a lot more palatable

I tried, it didn't work. The idea was to use `_Generic()` to issue the `TO_FLASH()` instead. The first issue is that `_Generic()` cannot tell string literal (which are the only type compatible with `TO_FLASH()`) apart from `char *` to something that is actually allocated in RAM. Here, the solution was to just force with coccinelle that the first parameter of `printf()` always a string literal or `__flash const char *`. I didn't really solve this yet, as I have to dive a bit deeper into the coccinelle doc to understand how this could be expressed as semantic patch.

The second issue however is a blocking one: The `TO_FLASH()` macro really only works at a preprocessor level. But `_Generic()` works at the C language level. So there is absolutely no way to do this with `_Generic()`. I am almost certain that with even the most recent C standard, there is no way to get rid of `TO_FLASH()`.

One could provide a `printf_literal(x, ...)` that would be an alias for `flash_printf(TO_FLASH(x), ...)`. But I think that this would actually increase the complexity without being a significant gain in readability.

benpicco commented:
> One could provide a `printf_literal(x, ...)` that would be an alias for `flash_printf(TO_FLASH(x), ...)`. But I think that this would actually increase the complexity without being a significant gain in readability.

Hm you think so? I feel like most `printf` invocations will have the format string as string literal, the cases where the format string is a variable (but still `const` on flash`) seem to be so rare that I would even not care about them at all.

Having that macro in every printf statement on the other hand is *very* ugly.

benpicco commented on sys/ps/ps.c:
```diff
@@ -43,31 +44,31 @@ void ps(void)
 #endif
 
-    printf("\tpid | "
+    flash_printf(TO_FLASH("\tpid | "
```
So this could be just
```suggestion
    printf("\tpid | "
```

now?

benpicco reviewed (approved):
The impact of this on non-AVR platforms is now minimal and the implicit `TO_FLASH` for `printf()` means this will do the right thing automatically in most of the cases.

Nice one, please squash!

maribu commented:
> Ok, CI still has a [different opinion](https://ci.riot-os.org/details/6b35c0debddc4617a6870ca7e4357c30) on this

The AVR libc has a define for the macro `AES_KEY`:

```
$ rg AES_KEY /usr/avr/include
[...]
/usr/avr/include/avr/iom256rfr2.h
4543:#define AES_KEY                         _SFR_MEM8(0x13F)
[...]
```

The AVR stdio wrapper needs to include `<avr/pgmspace.h>` on AVR to handle the shifting to progmem, but that in turn includes `<avr/io.h>` which for some AVR devices provides an `AES_KEY` macro :/

benpicco commented on cpu/avr8_common/include/stdio.h:
```diff
@@ -0,0 +1,122 @@
+ */
+#define printf(...)     /* implementation details */
+#else
+#define _TAKE_65TH_TOKEN( _1,  _2,  _3,  _4,  _5,  _6,  _7,  _8,  _9, _10, \
+                          _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, \
+                          _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, \
+                          _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, \
+                          _41, _42, _43, _44, _45, _46, _47, _48, _49, _50, \
+                          _51, _52, _53, _54, _55, _56, _57, _58, _59, _60, \
+                          _61, _62, _63, _64, N, ...) N
```
This could need some explanation

maribu replied:
Done. Is that fine now?

benpicco commented on sys/crypto/aes.c:
```diff
@@ -32,10 +32,6 @@
 
-#include <stdio.h>
-#include <stdarg.h>
-#include <string.h>
```
You still need `string.h` for `memset()`

maribu replied:
oops, yes. Sorry, I'm afk for two hours

benpicco commented:
I don't get why we have

```C
typedef struct aes_key_st {
    /** @cond INTERNAL */
    uint32_t rd_key[4 * (AES_MAXNR + 1)];
    int rounds;
    /** @endcond */
} AES_KEY;
```

RIOT coding convention would be `aes_key_t`

kfessel commented on cpu/avr8_common/include/stdio.h:
```diff
@@ -0,0 +1,122 @@
+#define _FPRINTF_SINGLEARG(stream, x) \
+        flash_fprintf(stream, TO_FLASH(x))
+#define _FPRINTF_MULTIARG(stream, x, ...) \
+        flash_printf(stream, TO_FLASH(x), __VA_ARGS__)
```
```suggestion
        flash_fprintf(stream, TO_FLASH(x), __VA_ARGS__)
```

maribu commented:
Agreed! But I'd rather not change this in this PR :)

kfessel commented on cpu/avr8_common/include/stdio.h:
```diff
@@ -0,0 +1,122 @@
+#define vfprintf(stream, fmt, args) flash_vprintf(stream, TO_FLASH(fmt), args)
+
+#define _SNPRINTF_SINGLEARG(buf, buf_len, fmt) \
+        flash_fprintf(buf, buf_len, TO_FLASH(fmt))
```
```suggestion
        flash_snprintf(buf, buf_len, TO_FLASH(fmt))
```

kfessel commented on cpu/avr8_common/include/stdio.h:
```diff
@@ -0,0 +1,122 @@
+#define _SNPRINTF_SINGLEARG(buf, buf_len, fmt) \
+        flash_fprintf(buf, buf_len, TO_FLASH(fmt))
+#define _SNPRINTF_MULTIARG(buf, buf_len, fmt, ...) \
+        flash_printf(buf, buf_len, TO_FLASH(fmt), __VA_ARGS__)
```
```suggestion
        flash_snprintf(buf, buf_len, TO_FLASH(fmt), __VA_ARGS__)
```

kfessel reviewed (commented):
some missing letters ?

maribu commented:
> some missing letters ?

Good catch, thx!

maribu commented:
> some missing letters ?

Good catch, thx!

maribu commented:
I squashed and dropped the changes to fix the naming clash with AES_KEY, as https://github.com/RIOT-OS/RIOT/pull/19290 is fixing that better and on its way in. Once that is merged, this PR hopefully passes the CI.

maribu commented:
@kaspar030: I think all your concerns have been addressed. Care to take another look?

fabian18 commented on sys/shell/shell.c:
```diff
@@ -99,8 +99,8 @@ static shell_command_handler_t search_commands_xfa(char *command)
     unsigned n = XFA_LEN(shell_command_t*, shell_commands_xfa);
```
```suggestion
    unsigned n = XFA_LEN(shell_command_xfa_t*, shell_commands_xfa);
```

fabian18 commented on sys/phydat/phydat_str.c:
```diff
@@ -92,75 +94,153 @@ void phydat_dump(phydat_t *data, uint8_t dim)
+    /* Yeah, this is as bad as it looks... The function is deprecated for this
+     * reason and it will only affect AVR users, for whom this is a good
+     * trade-off. */
+    static char buf[8];
```
"permille" and "undefined" would not fit I think

benpicco replied:
```suggestion
    static char buf[10];
```

fabian18 commented on sys/shell/shell.c:
```diff
@@ -134,15 +134,16 @@ static void print_commands_xfa(void)
 {
     unsigned n = XFA_LEN(shell_command_t*, shell_commands_xfa);
```
```suggestion
    unsigned n = XFA_LEN(shell_command_xfa_t*, shell_commands_xfa);
```

maribu commented:
@fabian18: Good catch, thx :)

benpicco commented on tests/senml_phydat/main.c:
```diff
@@ -114,7 +114,7 @@ void test_phydat_to_senml_float(void)
         phydat_to_senml_float(&res, &(value_tests[i].phydat), value_tests[i].dim);
 
-        DEBUG("Float: %" PRIi16 "e%" PRIi16 " %s -> %.f %s\n",
+        DEBUG("Float: %" PRIi16 "e%" PRIi16 " %" PRIsflash " -> %.f %s\n",
```
Looks like the include for that is missing

benpicco replied:
Also in other places where only this was added

benpicco commented on tests/driver_bmx055/main.c:
```diff
@@ -44,7 +44,7 @@ int main(void)
         while (dev) {
             int dim = saul_reg_read(dev, &res);
-            printf("\nDev: %s\tType: %s\n", dev->name,
+            printf("\nDev: %s\tType: %" PRIsflash "\n", dev->name,
```
include is still missing for this one

benpicco commented on sys/phydat/phydat_str.c:
```diff
@@ -92,75 +94,153 @@ void phydat_dump(phydat_t *data, uint8_t dim)
-        case UNIT_M3:       return "m^3";
-        case UNIT_G:        return "g";
-        case UNIT_DPS:      return "dps";
-        case UNIT_GR:       return "G";
```
this was (correctly) changed to 'g', but now the test script needs an update

maribu commented:
waiting for

- [ ] https://github.com/RIOT-OS/RIOT/pull/19292

maribu commented:
I updated the `phydat` code to no longer fix the unit symbol for gram, so it no longer depends on fixing the phydat unit.

Fixing phydat units seemed to be a quick fix, but turned into a larger endeavor. So I'd rather not have this PR depending on fixing a unit symbol.

benpicco commented:
> Fixing phydat units seemed to be a quick fix, but turned into a larger endeavor. 

eh, you could have the ACK by just adding two defines :wink:

kaspar030 commented:
> @kaspar030: I think all your concerns have been addressed. Care to take another look?

I don't think they have. We're still changing how to store things in flash, for which on all but avr, a simple "const" is enough. Now, not anymore. We'll carry this extra burden for all architectures, forever.

I don't think that we should have these architecture specific hacks throughout the code base.

Limiting this to saul, phydat, and shell might be viable. Can we maybe discuss this at the VMA? I'll add it as topic.

maribu commented:
bors try

benpicco commented:
bors try

benpicco commented:
bors cancel
bors try

maribu commented:
bors cancel
bors try

maribu commented:
bors cancel

maribu commented:
bors try

kfessel commented on drivers/include/saul.h:
```diff
@@ -315,10 +316,41 @@ int saul_read_notsup(const void *dev, phydat_t *dat);
+ *          it would write (regardless of @p max_size)
  */
-const char *saul_class_to_str(const uint8_t class_id);
+ssize_t saul_class_write(char *dest, size_t max_size, uint8_t class_id);
```
```suggestion
ssize_t saul_class_write(char *dest, const size_t max_size, const uint8_t class_id);
```
the `const` in front of `class_id` will allow to not remove the `const` in front of `claas_id` of `saul_class_to_str`

maribu commented:
🎉 This now passes a full Murdock run :)

benpicco commented on dist/tools/insufficient_memory/add_insufficient_memory_board.sh:
```diff
@@ -61,7 +61,12 @@ for app in ${APPLICATIONS}; do
+                  -e "Some feature requirements are blacklisted:" \
+                  -e "not supported.  Stop." \
+                  -e "let the build continue on expected errors by setting CONTINUE_ON_EXPECTED_ERRORS=1" \
+                  "$TMPFILE" > /dev/null; then
```
This seems unrelated

maribu replied:
A rebase fixed it :)

benpicco commented:
This needs a rebase

miri64 commented on sys/include/flash_utils.h:
```diff
@@ -0,0 +1,255 @@
+ * and modern modified Harvard architectures (e.g. ARM, ATtiny, ...) as well
+ * as von-Neumann machines.
+ *
+ * The intention is to limit in-tree use to small small number of modules that
```
Is this word duplication intentionally?

```suggestion
 * The intention is to limit in-tree use to a very small number of modules that
```

kaspar030 commented on sys/include/flash_utils.h:
```diff
@@ -0,0 +1,255 @@
+ * No module needs to be selected, this is a header-only implementation that
+ * is always available.
+ *
+ * # Porting Code to Use `flash_utils`
+ *
+ * This is mainly targeting applications developers to ease developing apps
+ * that work well on both legacy modified Harvard architectures (e.g. ATmega)
+ * and modern modified Harvard architectures (e.g. ARM, ATtiny, ...) as well
+ * as von-Neumann machines.
+ *
+ * The intention is to limit in-tree use to small small number of modules that
+ * yield the most "bang for the buck" and not leak the use of `flash_utils`
+ * through the API. Specifically, reverting to not using `flash_utils` should
+ * not be noticed by any user (unless looking at memory consumption).
```
Perfect, thanks!

benpicco commented:
Bors merge

maribu commented:
:tada: :woman_dancing: :man_dancing: :confetti_ball:

maribu commented:
bors wtf?
